Fixes an issue where log rotation would cut transaction in half

tinwelint opened the pull request:
The rotation done in PhysicalLogFile only flushed the channel,
i.e. bytes already written to it. It didn't tell the writer to
empty its buffer into the channel first. This could sometimes
result in rotation happening in the middle of a transaction.
All the bytes would be there, just split across files and be
more difficult to read later on if not using proper log version
bridging and would eventually cause problems when first part of
split transaction would be pruned.

This change will have transactions be fully written in the log
they started being written to before rotation happens, without
introducing any performance penalty of appending, forcing
or rotation.

Included in this PR is also a fix for having recovery use LogicalTransactionStore
so that it's able to read split transactions, this also simplifies and reduces
the recovery code since lots of special code for reading log files is
now deleted in favour of existing ways of doing that.

changelog: Fixes [#7204](https://github.com/neo4j/neo4j/issues/7204) an issue where log rotation sometimes would cut a transaction in half making it problematic for recovery to read such transactions. Transactions are now never split into multiple log files. Any db experiencing this issue can now also start that db with this fix included due to improved log reading abilities.

tinwelint committed:
Fixes an issue where log rotation would cut transaction in half

The rotation done in PhysicalLogFile only flushed the channel,
i.e. bytes already written to it. It didn't tell the writer to
empty its buffer into the channel first. This could sometimes
result in rotation happening in the middle of a transaction.
All the bytes would be there, just split across files and be
more difficult to read later on if not using proper log version
bridging and would eventually cause problems when first part of
split transaction would be pruned.

This change will have transactions be fully written in the log
they started being written to before rotation happens, without
introducing any performance penalty of appending, forcing
or rotation.

tinwelint committed:
Simplifies recovery code for reading transactions

essentially by using already existing means of reading transactions
from multiple files with seemless bridging instead of inventing
new and different ways of reading log file per log file.

This change has a couple of benefits:

- less code
- recovery algorithm can be seen in one place (Recovery class)
- ability to read transactions even if they would somehow cross
  log file boundaries.

There was an issue where log rotation may happen in the middle of
writing a transaction. Recovery would report weird log entries
because it read transactions from log files one by one in isolation
and fail to start database. That issues didn't corrupt the log files
and so starting such a database with this fix will have recovery
able to read those log files correctly.

tinwelint commented:
Please null-merge this or ping me about merging forward to 3.0 as it's slightly tricky and I've done it in a branch already. Thanks

tinwelint commented:
I've been running recovery-robustness on this branch for quite a while and it has been successful as well

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/LogicalTransactionStore.java:
```diff
@@ -23,25 +23,42 @@
+    /**
+     * Acquires a {@link TransactionCursor cursor} which will provide {@link CommittedTransactionRepresentation}
+     * instances for committed transactions, starting from the specified {@link LogPosition}.
+     * This is useful for placing a cursor at a position refered to by a {@link CheckPoint}.
```
referred

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/LogicalTransactionStore.java:
```diff
@@ -23,25 +23,42 @@
+     *
+     * @param position {@link LogPosition} of the first transaction that the cursor will return.
+     * @return an {@link TransactionCursor} capable of returning {@link CommittedTransactionRepresentation} instances
+     * for committed transactions, starting from the specified {@code transactionIdToStartFrom}.
```
transactionIdToStartFrom is no longer here :)

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/LogicalTransactionStore.java:
```diff
@@ -23,25 +23,42 @@
+     * or if the transaction has been committed, but information about it is no longer available for some reason.
+     * @throws IOException if there was an I/O related error looking for the start transaction.
+     */
+    TransactionCursor getTransactions( LogPosition position )
```
NoSuchTransactionException is a subclass of IOException and declaration of it is redundant

tinwelint replied:
Was following convention, perhaps it's declared to be explicit... just like IOException, ReadPastEndException (in log entry reading), which is very nice to advertise

MishaDemianenko replied:
that's maybe true except some of the implementations do that, some not probably depending who opens them last )

tinwelint replied:
Right, we can remove those from implementors as we see them (there was one mentioned below which is now fixed).

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -1030,31 +1017,33 @@ private void buildRecovery( final FileSystemAbstraction fileSystemAbstraction, C
             {
-                startupStatistics.setNumberOfRecoveredTransactions( recoveredCount.get() );
-                recoveredCount.set( 0 );
+                startupStatistics.setNumberOfRecoveredTransactions( numberOfRecoveredTransactions );
```
This is already great improvement to what it was before, but would it be much more easier pass startupStatistic inside recovery and update number of recovered transactions directly after migration without using monitors?

tinwelint replied:
Dunno, it's nice in a way to keep recovery as generic as possible. Don't think we should change more than required in a maint version this far back.

MishaDemianenko replied:
ok

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/PhysicalLogFile.java:
```diff
@@ -149,7 +149,13 @@ private PhysicalLogVersionedStoreChannel rotate( LogVersionedStoreChannel curren
+         * in the buffer for future flushing. Flushing that buffer now makes the last appended
+         * transaction complete in the log we're rotating away. Awesome.
+         */
+        writer.emptyBufferIntoChannelAndClearIt().flush();
```
would it make more sense if we change LogVersionedStoreChannel to perform buffer flush first as part of channel flush? That will prevent case when we will forget to call `emptyBufferIntoChannelAndClearIt`

tinwelint replied:
Yeah, that's a good alternative to this actually, will try it out

tinwelint replied:
Hmm, actually this turns out to be two separate things. Flushing can only be done by calling `emptyBufferIntoChannelAndClearIt()` unless you have a direct reference to the underlying channel, which is what rotation code in PhysicalLogFile did wrong.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/PhysicalLogicalTransactionStore.java:
```diff
@@ -49,7 +49,13 @@ public PhysicalLogicalTransactionStore( LogFile logFile, TransactionMetadataCach
 
     @Override
-    public IOCursor<CommittedTransactionRepresentation> getTransactions( final long transactionIdToStartFrom )
+    public TransactionCursor getTransactions( LogPosition position ) throws NoSuchTransactionException, IOException
```
NoSuchTransactionException is redundant here

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/recovery/DefaultRecoverySPI.java:
```diff
@@ -20,54 +20,62 @@
             RecoveryLegacyIndexApplierLookup legacyIndexApplierLookup,
             StoreFlusher storeFlusher, NeoStores neoStores,
-            Visitor<LogVersionedStoreChannel,IOException> logFileRecoverer,
             PhysicalLogFiles logFiles, FileSystemAbstraction fileSystemAbstraction,
```
logFiles and fileSystemAbstraction is no longer used and should be removed

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/recovery/DefaultRecoverySPI.java:
```diff
@@ -88,61 +96,64 @@ public void forceEverything()
+                lastRecoveredTransaction.getCommitEntry().getTimeWritten(),
+                positionAfterLastRecoveredTransaction.getByteOffset(),
+                positionAfterLastRecoveredTransaction.getLogVersion() );
+        // TODO: Also truncate last log after last known position
```
TODO broke through :)

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/recovery/DefaultRecoverySPI.java:
```diff
@@ -20,54 +20,62 @@
+import static org.neo4j.kernel.impl.api.TransactionApplicationMode.RECOVERY;
 
-public class DefaultRecoverySPI implements Recovery.SPI
+public class DefaultRecoverySPI implements Recovery.SPI, Visitor<CommittedTransactionRepresentation,Exception>
```
I would say that current design where recovery spi is also a visitor i find  more overloaded then previous version where those a 2 different things

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/recovery/DefaultRecoverySPI.java:
```diff
@@ -88,61 +96,64 @@ public void forceEverything()
+        // each store is aware that recovery will be performed. At this point all the stores have
+        // already started btw.
+        // Go and read more at {@link CommonAbstractStore#deleteIdGenerator()}
+        neoStores.deleteIdGenerators();
```
this looks strange now....critical call in a getter...

tinwelint replied:
I wanted to remove the constraint that the `recoveryRequired()` needed to be called before `getRecoveryVisitor()`. Renaming `getRecoveryVisitor()` to `startRecovery()` feels better, wdyt?

MishaDemianenko replied:
yes sounds much better

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/log/TransactionLogAppendAndRotateStressIT.java:
```diff
@@ -0,0 +1,276 @@
+import static org.neo4j.kernel.impl.transaction.tracing.LogAppendEvent.NULL;
+import static org.neo4j.kernel.impl.util.IdOrderingQueue.BYPASS;
+
+public class TransactionLogAppendAndRotateStressIT
```
should we create separate stress test build for it instead?

tinwelint replied:
It runs in 2 seconds or so, perhaps it could simply be an IT?

MishaDemianenko replied:
ok then lets to it simple IT

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/log/TransactionLogAppendAndRotateStressIT.java:
```diff
@@ -0,0 +1,276 @@
+                        }
+                        catch ( Exception e )
+                        {
+                            e.printStackTrace( System.out );
```
SuppressOutput rule instead?

tinwelint replied:
That wouldn't help because any exception fails the test and so SuppressOutput would print it anyway. Having this printing should be fine since it won't print on green test and you want to have more information when failing.

MishaDemianenko replied:
ok fair

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/log/TransactionLogAppendAndRotateStressIT.java:
```diff
@@ -0,0 +1,276 @@
+                {
+                case 0: // Expects start entry
+                    assertTrue( entry instanceof LogEntryStart );
+                    state = 1;
```
what if we will do state++ here

tinwelint replied:
It can be simplified with a boolean instead, since it's either 0 or 1

MishaDemianenko replied:
that's true but with counter we will check that number of entry start is the same as number of commits

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/log/TransactionLogAppendAndRotateStressIT.java:
```diff
@@ -0,0 +1,276 @@
+                    assertTrue( entry instanceof LogEntryCommand || entry instanceof LogEntryCommit );
+                    if ( entry instanceof LogEntryCommit )
+                    {
+                        state = 0;
```
and state -- here

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/log/TransactionLogAppendAndRotateStressIT.java:
```diff
@@ -0,0 +1,276 @@
+                default: throw new IllegalArgumentException();
+                }
+            }
+            assertEquals( 0, state );
```
and final state should be 0 here as result :) ?

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/RecoveryTest.java:
```diff
@@ -275,6 +275,10 @@ public void recoveryRequired()
         }
     }
 
+    // TODO: Test about truncate file
```
TODO's

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/PhysicalTransactionCursor.java:
```diff
@@ -33,18 +33,18 @@
 
 public class PhysicalTransactionCursor<T extends ReadableLogChannel>
-        implements IOCursor<CommittedTransactionRepresentation>
+        implements TransactionCursor
```
there is `protected List<Command> commandList()` which can be inlined now

tinwelint replied:
fixed

MishaDemianenko commented:
@tinwelint review round completed, couple of comments.
It looks great and much more simple then it was before!

tinwelint commented on community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/log/TransactionLogAppendAndRotateStressIT.java:
```diff
@@ -0,0 +1,276 @@
+                        }
+                        catch ( Exception e )
+                        {
+                            e.printStackTrace( System.out );
```
I think we can remove this printStackTrace since a test failure will show the exception anyway.

tinwelint committed:
Further simplifications of recovery code

and some comment typos fixed

tinwelint committed:
Separated recovery visitor from DefaultRecoverySPI

and puts more weight on getRecoveryVisitor() method by renaming to
startRecovery(). This way it makes more sense to add more responsibilities to it.

tinwelint committed:
Implemented truncation of last log after recovery

which was sitting as a TODO after previous change to recovery code

tinwelint committed:
Changed name from StressIT to IT so that it runs with full build

tinwelint commented:
@MishaDemianenko added a couple of more commits addressing the comments

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/RecoveryTest.java:
```diff
@@ -275,6 +275,154 @@ public void recoveryRequired()
+        LifeSupport life = new LifeSupport();
+        Recovery.Monitor monitor = mock( Recovery.Monitor.class );
+        final AtomicBoolean recoveryRequired = new AtomicBoolean();
+        try
```
this whole try catch block duplicates and should be extracted

MishaDemianenko commented:
@tinwelint small comment about test, otherwise ready to merge

tinwelint committed:
Deduplicated test code block
